### DISCUSSION TITLE
Nextcloud stops responding/crashes when browsing images

### DISCUSSION THREAD

**Discussion by chefexperte** [nextcloud/all-in-one#3288] [Questions] [answered]
### Steps to reproduce
1. Start Nextcloud
2. Browse images, preferrably a folder with many images
3. Wait for few previews to load, then click on an image. 
4. Nexcloud stops responding completely.

### Expected behavior 
When browsing images, the previews load and show the images. 

### Actual behavior 
The server becomes unresponsive, and nothing works anymore. First I get 100% CPU usage, then Nextcloud stops responding completely, except for the AIO interface. 

### Host OS 
I am using Debian 12 (bookworm)

#### Nextcloud AIO version 
Nextcloud AIO v7.0.0

#### Current channel 
*latest* channel

#### Other valuable info 
Nextcloud does not crash in a classical sense, but becomes totally unresponsive. 

Nextcloud container starts throwing 
`WARNING: [pool www] server reached max_children setting (158), consider raising it` 
after a few minutes. 

And at some point I also got a 
<details>
<summary>
Nextcloud container log
</summary>

```
Doctrine\DBAL\Exception: Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] connection to server at "nextcloud-aio-database" (172.21.0.5), port 5432 failed: FATAL:  remaining connection slots are reserved for non-replication superuser connections in /var/www/html/lib/private/DB/Connection.php:140

ace:
#0 /var/www/html/3rdparty/doctrine/dbal/src/Connection.php(1531): OC\DB\Connection->connect()
#1 /var/www/html/3rdparty/doctrine/dbal/src/Connection.php(1029): Doctrine\DBAL\Connection->getWrappedConnection()
#2 /var/www/html/lib/private/DB/Connection.php(262): Doctrine\DBAL\Connection->executeQuery('SELECT * FROM "...', Array, Array, NULL)
#3 /var/www/html/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php(345): OC\DB\Connection->executeQuery('SELECT * FROM "...', Array, Array)
#4 /var/www/html/lib/private/DB/QueryBuilder/QueryBuilder.php(280): Doctrine\DBAL\Query\QueryBuilder->execute()
#5 /var/www/html/lib/private/AppConfig.php(418): OC\DB\QueryBuilder\QueryBuilder->execute()
#6 /var/www/html/lib/private/AppConfig.php(184): OC\AppConfig->loadConfigValues()
#7 /var/www/html/lib/private/AppConfig.php(374): OC\AppConfig->getApps()
#8 /var/www/html/lib/private/legacy/OC_App.php(803): OC\AppConfig->getValues(false, 'installed_versi...')
#9 /var/www/html/lib/private/Server.php(733): OC_App::getAppVersions()
#10 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\Server->OC\{closure}(Object(OC\Server))
#11 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#12 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OC\\Memcache\\Fac...')
#13 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OC\\Memcache\\Fac...', true)
#14 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OC\\Memcache\\Fac...')
#15 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(193): OC\AppFramework\Utility\SimpleContainer->get('OC\\Memcache\\Fac...')
#16 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(OC\Server))
#17 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(118): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#18 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OCP\\ICacheFacto...')
#19 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OCP\\ICacheFacto...', true)
#20 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OCP\\ICacheFacto...')
#21 /var/www/html/lib/private/Server.php(1115): OC\AppFramework\Utility\SimpleContainer->get('OCP\\ICacheFacto...')
#22 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\Server->OC\{closure}(Object(OC\Server))
#23 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#24 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OCP\\Lock\\ILocki...')
#25 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OCP\\Lock\\ILocki...', true)
#26 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OCP\\Lock\\ILocki...')
#27 /var/www/html/lib/private/Server.php(2067): OC\AppFramework\Utility\SimpleContainer->get('OCP\\Lock\\ILocki...')
#28 /var/www/html/lib/private/Files/View.php(106): OC\Server->getLockingProvider()
#29 /var/www/html/lib/private/Server.php(467): OC\Files\View->__construct()
#30 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\Server->OC\{closure}(Object(OC\Server))
#31 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#32 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OC\\Files\\Node\\H...')
#33 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OC\\Files\\Node\\H...', true)
#34 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OC\\Files\\Node\\H...')
#35 /var/www/html/lib/private/Server.php(1469): OC\AppFramework\Utility\SimpleContainer->get('OC\\Files\\Node\\H...')
#36 /var/www/html/lib/base.php(638): OC\Server->boot()
#37 /var/www/html/lib/base.php(1180): OC::init()
#38 /var/www/html/cron.php(43): require_once('/var/www/html/l...')
#39 {main}
```
</details>

Which goes hand in hand with the Database log 

`FATAL:  remaining connection slots are reserved for non-replication superuser connections`

I think this is a problem with imaginary, it spawns lots of imaginary processes when browsing images, which cause the 100% CPU usage, and the problem only happens when browsing images. 

The imaginary container however does not throw any weird errors, it just does the normal 
`172.21.0.8 - - [29/Aug/2023 14:05:56] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2280%22%7D%7D%5D HTTP/1.1" 200 13462 1.3307`
stuff.

**Comment by szaimen** [comment]
Hi, can you for a test disable collabora from the AIO interface and check if this improves things?

**Reply by mangabs** [reply]
Hi all,

I have the same problem. I am unable today to get a working production environment and this is very annoying.
I made multiple tests on Debian 12: Nextcloud AIO v7.0.0 with default options (Collabora, Imaginary, Nextcloud Talk).
@szaimen : Yes, if I disable Collabora from the AIO interface everything is working properly. Image previews loads well and no freeze on Nextcloud web interface.

**Comment by Chouffy** [comment]
Hi all, I have the same symptoms however steps to reproduce doesn't apply to me - it just hangs sometime.
This _could_ be caused by the regular scan every 15 min using [PreviewGenerator](https://github.com/rullzer/previewgenerator)

@szaimen I've disable Collabora (and ClamAV) to be sure, the only addon left is Imaginary
I'll monitor the situation and update in few days

**Reply by Chouffy** [reply]
My instance runs fine since 5 days after I "monitor" (= check on HTTP code 200) the situation, even with ClamAV, Collabora, Imaginary and PreviewGenerator activated … so this issue is just "gone" 🤔

**Reply by Chouffy** [reply]
Well, after a cold restart of my server, the issue just came back exactly as described... 😢

Situation: "standard" nextcloud AIO with Collabora, ClamAV, Imaginary

Below are the logs - @szaimen do you think you could have a minute to look at them? thanks a lot in advance
[collabora.txt](https://github.com/nextcloud/all-in-one/files/12548574/collabora.txt)
[apache.txt](https://github.com/nextcloud/all-in-one/files/12548575/apache.txt)
[imaginary.txt](https://github.com/nextcloud/all-in-one/files/12548576/imaginary.txt)
[clamav.txt](https://github.com/nextcloud/all-in-one/files/12548577/clamav.txt)
[redis.txt](https://github.com/nextcloud/all-in-one/files/12548578/redis.txt)
[notify-push.txt](https://github.com/nextcloud/all-in-one/files/12548579/notify-push.txt)
[nextcloud.txt](https://github.com/nextcloud/all-in-one/files/12548580/nextcloud.txt)
[database.txt](https://github.com/nextcloud/all-in-one/files/12548581/database.txt)

After disabling all addon (Collabora, ClamAV, Imaginary), it works, but there are still errors in the logs (!) … but nextcloud didn't crashed - maybe it was just close to:

[apache_2.txt](https://github.com/nextcloud/all-in-one/files/12548645/apache_2.txt)
[database_2.txt](https://github.com/nextcloud/all-in-one/files/12548646/database_2.txt)
[nextcloud_2.txt](https://github.com/nextcloud/all-in-one/files/12548647/nextcloud_2.txt)
[notify-push_2.txt](https://github.com/nextcloud/all-in-one/files/12548648/notify-push_2.txt)
[redis_2.txt](https://github.com/nextcloud/all-in-one/files/12548649/redis_2.txt)

**Reply by szaimen** [reply]
Disabling collabora is the correct workaround until a new richdocuments app version is published which contains https://github.com/nextcloud/richdocuments/pull/3146

**Comment by szaimen** [comment]
Mlving to discussions while waiting

**Comment by imfbsbn** [comment]
I have a similar situation – forgive me if this is already been covered, I have searched, this is the best discussion I could find.

Installed AIO on Linode. Ubuntu 22.04 LTS.

Have not browsed any images. However when working with files logging shows:
   RuntimeException: Unable to read stream contents: Automatic conversion of false to array is deprecated

First symptom is that pages, particularly the dashboard, will never finish loading. The page loads and appears to be complete. But the browser continues to wait – the circle never stops spinning. CPU usage gets a little squirrelly. Bounces around and settles down once the user presses escape to stop the page from continuing to try and load.

After this error occurs three or four times NC simply stops responding. Have to reboot the server.

Total noob. Any guidance would be much appreciated.

**Reply by szaimen** [reply]
> RuntimeException: Unable to read stream contents: Automatic conversion of false to array is deprecated

This issue is unrelated to this thread.

**Comment by geevarghesest** [comment]
Hi,
i'm facing the same error `FATAL:  remaining connection slots are reserved for non-replication superuser connections`
with AIO v7.2.1 ( nextcloud 27.1.0 ).
Disabling collbora alone didn't solve the issue. After disabling imaginary too, the problem seems to be solved

**Comment by mangabs** [comment]
Hi,
I build a new instance version (7.4.1) with default options (Collabora, Imaginary, Nextcloud Talk). The problem is still there: no image preview and this message in nextcloud-aio-database: "FATAL:  remaining connection slots are reserved for non-replication superuser connections".

Could you please tell us if ticket is open for this problem. I can't find one (in github AIO? in github RichDocument? in Imaginary?)
https://github.com/nextcloud/richdocuments/pull/3146 seems not to be sufficient to solve the issue.

**Comment by MarvvanPal** [comment]
I added my experience to this topic. See #2839

**Comment by szaimen** [comment]
Hi, we made some improvements to the config with v7.6.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Comment by szaimen** [comment, marked as answer]
Hi, I found the actual problem finally: https://github.com/nextcloud/server/pull/41263. It will be most likely [included in 27.1.4](https://github.com/nextcloud/server/pull/41289)

**Reply by szaimen** [reply]
This is now released with v7.7.1 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
